distributed/schedulerstate.py

```diff
@@ -0,0 +1,1419 @@
+       Usually this is empty unless the task has been specifically restricted
+       to only run on certain hosts.  A hostname may correspond to one or
+       several connected workers.
+
+    .. attribute:: worker_restrictions: {worker addresses}
+
+       A set of complete worker addresses where this can be run (or ``None``
+       if empty).  Usually this is empty unless the task has been specifically
+       restricted to only run on certain workers.
+
+       Note this is tracking worker addresses, not worker states, since
+       the specific workers may not be connected at this time.
+
+    .. attribute:: resource_restrictions: {resource: quantity}
+
+       Resources required by this task, such as ``{'gpu': 1}`` or
+       ``{'memory': 1e9}`` (or ``None`` if empty).  These are user-defined
+       names and are matched against the contents of each
+       :attr:`WorkerState.resources` dictionary.
+
+    .. attribute:: loose_restrictions: bool
+
+       If ``False``, each of :attr:`host_restrictions`,
+       :attr:`worker_restrictions` and :attr:`resource_restrictions` is
+       a hard constraint: if no worker is available satisfying those
+       restrictions, the task cannot go into the "processing" state and
+       will instead go into the "no-worker" state.
+
+       If ``True``, the above restrictions are mere preferences: if no worker
+       is available satisfying those restrictions, the task can still go
+       into the "processing" state and be sent for execution to another
+       connected worker.
+
+    """
+
+    __slots__ = (
+        # === General description ===
+        # Key name
+        'key',
+        # Key prefix (see key_split())
+        'prefix',
+        # How to run the task (None if pure data)
+        'run_spec',
+        # Alive dependents and dependencies
+        'dependencies',
+        'dependents',
+        # Compute priority
+        'priority',
+        # Restrictions
+        'host_restrictions',
+        'worker_restrictions',  # not WorkerStates but addresses
+        'resource_restrictions',
+        'loose_restrictions',
+        # === Task state ===
+        'state',
+        # Whether some dependencies were forgotten
+        'has_lost_dependencies',
+        # If in 'waiting' state, which tasks need to complete
+        # before we can run
+        'waiting_on',
+        # If in 'waiting' or 'processing' state, which tasks needs us
+        # to complete before they can run
+        'waiters',
+        # In in 'processing' state, which worker we are processing on
+        'processing_on',
+        # If in 'memory' state, Which workers have us
+        'who_has',
+        # Which clients want us
+        'who_wants',
+        'exception',
+        'traceback',
+        'exception_blame',
+        'suspicious',
+        'retries',
+        'nbytes',
+        )
+    if cython.compiled:
+        key = cython.declare(object, visibility='public')
```

Comment:
You can also spell this `cython.declare(key=object, visibility='public')`. Avoids an assignment and might be more convenient.